Midichlorians/Server/server.h

```diff
@@ -0,0 +1,229 @@
+    try {
+      ios_event = ParseJSON<Variant<T_IOS_EVENTS>>(line);
+    } catch (const InvalidJSONException&) {
+      return false;
+    }
+
+    PassLogEntryToConsumer(EventLogEntry(current::time::Now(), std::move(ios_event)));
+    return true;
+  }
+
+  bool RequestAsWebEvent(const Request& r) {
+    using namespace current::midichlorians::web;
+
+    std::map<std::string, std::string> post_q;  // Query parameters extracted from POST body.
+    const std::map<std::string, std::string>* q;
+    const std::map<std::string, std::string>* h = &r.headers;
+
+    if (r.method == "GET" || r.method == "HEAD") {
+      q = &r.url.AllQueryParameters();
+    } else if (r.method == "POST") {
+      post_q = blocks::impl::URLParametersExtractor("?" + r.body).AllQueryParameters();
+      for (const auto& cit : r.url.AllQueryParameters()) {
+        post_q.insert(cit);
+      }
+      q = &post_q;
+    } else {
+      return false;
+    }
+
+    Variant<T_WEB_EVENTS> web_event;
+    try {
+      if (q->at("ea") == "En") {
+        web_event = WebEnterEvent();
+      } else if (q->at("ea") == "Ex") {
+        web_event = WebExitEvent();
+      } else if (q->at("ea") == "Fg") {
+        web_event = WebForegroundEvent();
+      } else if (q->at("ea") == "Bg") {
+        web_event = WebBackgroundEvent();
+      } else {
+        web_event = WebGenericEvent(q->at("ec"), q->at("ea"));
+      }
+    } catch (const std::out_of_range&) {
+      return false;
+    }
+
+    if (!ExtractWebBaseEventFields(q, h, web_event)) {
+      return false;
+    }
+
+    PassLogEntryToConsumer(EventLogEntry(current::time::Now(), std::move(web_event)));
+    return true;
+  }
+
+  template <typename T_LOG_ENTRY>
+  void PassLogEntryToConsumer(T_LOG_ENTRY&& entry, bool is_valid_entry = true) {
+    T_LOG_ENTRY_VARIANT entry_variant(std::move(entry));
+    entry_variant.Call(log_entry_consumer_);
+    if (is_valid_entry) {
+      ++events_pushed_;
+      last_event_t_ = Value<LogEntryBase>(entry_variant).server_us;
+    }
+  }
+
+  bool ExtractWebBaseEventFields(const std::map<std::string, std::string>* q,
+                                 const std::map<std::string, std::string>* h,
+                                 Variant<T_WEB_EVENTS>& web_event) {
+    using namespace current::midichlorians::web;
+
+    assert(q);
+    assert(h);
+
+    WebBaseEvent& dest_event = Value<WebBaseEvent>(web_event);
+    try {
+      dest_event.user_ms = std::chrono::milliseconds(strings::FromString<uint64_t>(q->at("_t")));
+      dest_event.customer_id = q->at("CUSTOMER_ACCOUNT");
+      dest_event.client_id = q->at("cid");
+      dest_event.ip = h->at("X-Forwarded-For");
+      dest_event.user_agent = h->at("User-Agent");
+      blocks::URL url = blocks::URL(h->at("Referer"));
+      dest_event.referer_host = url.host;
+      dest_event.referer_path = url.path;
+      dest_event.referer_querystring = url.query.AsImmutableMap();
+    } catch (const std::out_of_range&) {
+      return false;
```

Comment:
It's probably a good idea to still process an event if most of the fields are present. Worth a quick chat tomorrow.